ProjectLighthouse/Helpers/CensorHelper.cs

```diff
@@ -19,80 +21,82 @@ public static class CensorHelper
         if (CensorConfiguration.Instance.UserInputFilterMode == FilterMode.None) return message;
 
-        int profaneIndex;
+        int profaneCount = 0;
 
         foreach (string profanity in CensorConfiguration.Instance.FilteredWordList)
+        {
+            int lastFoundProfanity = 0;
+            int profaneIndex;
+            List<int> censorIndices = new List<int>();
+            
             do
             {
-                profaneIndex = message.ToLower().IndexOf(profanity, StringComparison.Ordinal);
-                if (profaneIndex != -1) message = Censor(profaneIndex, profanity.Length, message);
+                profaneIndex = message.IndexOf(profanity, lastFoundProfanity, StringComparison.OrdinalIgnoreCase);
+
+                if (profaneIndex == -1) continue;
+
+                censorIndices.Add(profaneIndex);
+                
+                lastFoundProfanity = profaneIndex + profanity.Length;
             }
             while (profaneIndex != -1);
 
-        return message;
-    }
+            for (int i = censorIndices.Count - 1; i >= 0; i--)
+            {
+                Censor(censorIndices[i], profanity.Length, stringBuilder);
+                profaneCount += 1;
+            }
+        }
 
-    private static string Censor(int profanityIndex, int profanityLength, string message)
-    {
-        StringBuilder sb = new();
+        if (profaneCount > 0 && message.Length <= 94 && ServerConfiguration.Instance.LogChatFiltering) // 94 = lbp char limit
+            Logger.Info($"Censored {profaneCount} profane words from message \"{message}\"", LogArea.Filter);
 
-        sb.Append(message.AsSpan(0, profanityIndex));
+        return stringBuilder.ToString();
+    }
 
+    private static void Censor(int profanityIndex, int profanityLength, StringBuilder message)
+    {
         switch (CensorConfiguration.Instance.UserInputFilterMode)
         {
             case FilterMode.Random:
                 char prevRandomChar = '\0';
-                for (int i = 0; i < profanityLength; i++)
+                for (int i = profanityIndex; i < profanityIndex + profanityLength; i++)
                 {
-                    if (message[i] == ' ')
-                    {
-                        sb.Append(' ');
-                    }
-                    else
-                    {
-                        char randomChar = randomCharacters[CryptoHelper.GenerateRandomInt32(0, randomCharacters.Length)];
-                        if (randomChar == prevRandomChar) randomChar = randomCharacters[CryptoHelper.GenerateRandomInt32(0, randomCharacters.Length)];
-
-                        prevRandomChar = randomChar;
-                        sb.Append(randomChar);
-                    }
+                    if (char.IsWhiteSpace(message[i])) continue;
+                    
+                    char randomChar = randomCharacters[CryptoHelper.GenerateRandomInt32(0, randomCharacters.Length)];
+                    if (randomChar == prevRandomChar)
```

Comment:
There is a non-zero chance that this could run forever but it guarantees that no two neighboring characters are identical
```suggestion
                    while (randomChar == prevRandomChar)
```